File: backend/conversation/session_manager.py
"""
Darukaa.Earth AI Biodiversity Intelligence System
Session State & Multi-Turn Memory Manager
Persists user conversational context, extracted parameters, and dialogue turns.
"""

from typing import Dict, Any, List, Optional
import time
import uuid

class ConversationSession:
    def __init__(self, session_id: str):
        self.session_id = session_id
        self.created_at = time.time()
        self.last_active = time.time()
        self.messages: List[Dict[str, Any]] = []
        # Accumulated ecological profile
        self.profile: Dict[str, Any] = {
            "soc_pct": None,
            "rainfall": None,
            "land_use": None,
            "region": None,
            "soil_ph": None,
            "tillage": None,
            "pesticides": None,
            "latitude": None,
            "longitude": None,
            "notes": []
        }
        self.clarification_requested: bool = False
        self.pending_variables: List[str] = []

    def add_message(self, role: str, content: str, metadata: Optional[Dict[str, Any]] = None):
        self.messages.append({
            "id": str(uuid.uuid4()),
            "role": role,
            "content": content,
            "timestamp": time.time(),
            "metadata": metadata or {}
        })
        self.last_active = time.time()

    def update_profile(self, new_params: Dict[str, Any]):
        for k, v in new_params.items():
            if v is not None and v != "":
                self.profile[k] = v
        self.last_active = time.time()

    def get_missing_critical_variables(self) -> List[str]:
        missing = []
        if self.profile.get("soc_pct") is None:
            missing.append("soil organic carbon %")
        if not self.profile.get("rainfall") and not self.profile.get("region"):
            missing.append("rainfall pattern / climate region")
        if not self.profile.get("land_use"):
            missing.append("current land use / crop type")
        return missing

    def is_complete_enough(self) -> bool:
        # Require at least 2 key variables (e.g., SOC + (rainfall or land_use)) or explicit diagnostic command
        provided_count = sum(1 for v in [self.profile.get("soc_pct"), self.profile.get("rainfall"), self.profile.get("land_use"), self.profile.get("region")] if v is not None)
        return provided_count >= 2

class SessionRegistry:
    def __init__(self):
        self.sessions: Dict[str, ConversationSession] = {}

    def get_or_create(self, session_id: Optional[str] = None) -> ConversationSession:
        if not session_id or session_id not in self.sessions:
            new_id = session_id or str(uuid.uuid4())
            self.sessions[new_id] = ConversationSession(new_id)
            return self.sessions[new_id]
        return self.sessions[session_id]

# Global Singleton session registry
session_registry = SessionRegistry()

File: backend/src/conversation/sessionManager.js
/**
 * Darukaa.Earth AI Biodiversity Intelligence System
 * Session Manager & Multi-Turn State Accumulator (ES Modules)
 */

export class ConversationSession {
  constructor(sessionId) {
    this.sessionId = sessionId;
    this.createdAt = Date.now();
    this.lastActive = Date.now();
    this.messages = [];
    this.profile = {
      soc_pct: null,
      rainfall: null,
      land_use: null,
      region: null,
      soil_ph: null,
      tillage: null,
      latitude: null,
      longitude: null
    };
    this.clarificationRequested = false;
    this.pendingVariables = [];
  }

  addMessage(role, content, metadata = {}) {
    this.messages.push({
      id: `${Date.now()}-${Math.random().toString(36).substring(2, 8)}`,
      role,
      content,
      timestamp: Date.now(),
      metadata
    });
    this.lastActive = Date.now();
  }

  updateProfile(params = {}) {
    Object.keys(params).forEach(key => {
      if (params[key] !== null && params[key] !== undefined && params[key] !== "") {
        this.profile[key] = params[key];
      }
    });
    this.lastActive = Date.now();
  }

  getMissingCriticalVariables() {
    const missing = [];
    if (this.profile.soc_pct === null) {
      missing.push("soil organic carbon %");
    }
    if (!this.profile.rainfall && !this.profile.region) {
      missing.push("rainfall pattern / climate region");
    }
    if (!this.profile.land_use) {
      missing.push("current land use / crop type");
    }
    return missing;
  }

  isCompleteEnough() {
    const count = [this.profile.soc_pct, this.profile.rainfall, this.profile.land_use, this.profile.region]
      .filter(v => v !== null && v !== undefined && v !== "").length;
    return count >= 2;
  }
}

export class SessionRegistry {
  constructor() {
    this.sessions = new Map();
  }

  getOrCreate(sessionId) {
    const id = sessionId || `session-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`;
    if (!this.sessions.has(id)) {
      this.sessions.set(id, new ConversationSession(id));
    }
    return this.sessions.get(id);
  }
}

export const sessionRegistry = new SessionRegistry();
